### ISSUE TITLE
MacOSX: CFAbsoluteTimeGetCurrent can go backwards triggering assertion.

### ISSUE THREAD

**Issue by lfnoise** [ocornut/imgui#4557]
Version:  1.85 WIP
Branch: master

Back-ends: imgui_impl_osx, imgui_impl_metal
Compiler: Xcode 12.5.1
Operating System: MacOS X 11.5.2

After leaving an imgui app running for a couple of days I found it crashed :

Assertion failed: ((g.IO.DeltaTime > 0.0f || g.FrameCount == 0) && "Need a positive DeltaTime!"), function ErrorCheckNewFrameSanityChecks, file /.../Imgui/imgui.cpp, line 7253.

ImGui_ImplOSX_NewFrame has this code that relies on CFAbsoluteTimeGetCurrent:

```
    // Setup time step
    if (g_Time == 0.0)
        g_Time = CFAbsoluteTimeGetCurrent();
    CFAbsoluteTime current_time = CFAbsoluteTimeGetCurrent();
    io.DeltaTime = (float)(current_time - g_Time);
    g_Time = current_time;
```

According to the docs for CFAbsoluteTimeGetCurrent(), "The system time may decrease due to synchronization with external time references or due to an explicit user change of the clock."
(I did not change my clock manually.)

Since CFAbsoluteTimeGetCurrent is not monotonic, imgui_impl_osx.mm should use a different time reference such as mach_absolute_time. Also probably the assertion should be made when DeltaTime is set, not when it is accessed.

**Comment by ocornut** [comment]
Hello,
Thanks for posting this. Could you provide a PR with a fix for it, using a suitable clock source?

**Comment by lfnoise** [comment]
I'm having security issues with github at the moment. I'll get it pushed when I can. In the meantime, here's the diff.
```
diff --git a/backends/imgui_impl_osx.mm b/backends/imgui_impl_osx.mm
index 1eb79d99..d0089352 100644
--- a/backends/imgui_impl_osx.mm
+++ b/backends/imgui_impl_osx.mm
@@ -16,6 +16,7 @@
 #include "imgui.h"
 #include "imgui_impl_osx.h"
 #import <Cocoa/Cocoa.h>
+#include <mach/mach_time.h>
 
 // CHANGELOG
 // (minor and older changes stripped away, please see git history for details)
@@ -37,7 +38,8 @@
 @class ImFocusObserver;
 
 // Data
-static CFAbsoluteTime g_Time = 0.0;
+static double         g_HostClockPeriod;
+static double         g_Time = 0.0;
 static NSCursor*      g_MouseCursors[ImGuiMouseCursor_COUNT] = {};
 static bool           g_MouseCursorHidden = false;
 static bool           g_MouseJustPressed[ImGuiMouseButton_COUNT] = {};
@@ -52,6 +54,19 @@ static ImFocusObserver* g_FocusObserver = NULL;
 + (id)_windowResizeEastWestCursor;
 @end
 
+static void InitHostClockPeriod()
+{
+    struct mach_timebase_info info;
+    mach_timebase_info(&info);
+    // Period is the reciprocal of frequency.
+    g_HostClockPeriod = 1e-9 * ((double)info.denom / (double)info.numer);
+}
+
+static double GetMachAbsoluteTimeInSeconds()
+{
+    return (double)mach_absolute_time() * g_HostClockPeriod;
+}
+
 static void resetKeys()
 {
     ImGuiIO& io = ImGui::GetIO();
@@ -231,9 +246,11 @@ void ImGui_ImplOSX_NewFrame(NSView* view)
     }
 
     // Setup time step
-    if (g_Time == 0.0)
-        g_Time = CFAbsoluteTimeGetCurrent();
-    CFAbsoluteTime current_time = CFAbsoluteTimeGetCurrent();
+    if (g_Time == 0.0) {
+        InitHostClockPeriod();
+        g_Time = GetMachAbsoluteTimeInSeconds();
+    }
+    double current_time = GetMachAbsoluteTimeInSeconds();
     io.DeltaTime = (float)(current_time - g_Time);
     g_Time = current_time;
 
```

**Comment by ocornut** [comment]
Merged as bc3d267c5, thank you very much!
